tests/test_main.py

```diff
@@ -0,0 +1,255 @@
+    dispatcher = Dispatcher([], groups)
+    with patch.object(argparse.ArgumentParser, 'format_help', str):
+        generated_help = dispatcher.main_parser.format_help()
+    return generated_help
+
+
+def test_customhelp_simplest():
+    """Simplest situation: one command in (of course) one group."""
+    groups = [
+        ('test-group', "simple stuff (test)", [create_command('cmd-name', 'cmd help blah')]),
+    ]
+    generated_help = get_generated_help(groups)
+    expected_help = """
+        commands:
+
+            simple stuff (test):
+                cmd-name   cmd help blah
+    """
+    assert generated_help == textwrap.dedent(expected_help)
+
+
+def test_customhelp_onegroup_severalcommands():
+    """Multiple commands in the same group.
+
+    Note that commands are ordered.
+    """
+    groups = [
+        ('test-group', 'whatever title', [
+            create_command('name1', 'cmd help 1'),
+            create_command('zz-name2', 'cmd help 2'),
+            create_command('other-name', 'super long help'),
+        ]),
+    ]
+    generated_help = get_generated_help(groups)
+    expected_help = """
+        commands:
+
+            whatever title:
+                name1        cmd help 1
+                other-name   super long help
+                zz-name2     cmd help 2
+    """
+    assert generated_help == textwrap.dedent(expected_help)
+
+
+def test_customhelp_several_simple_groups():
+    """Multiple groups (each one simple).
+
+    Note that the groups order are respected.
+    """
+    groups = [
+        ('test-group-1', 'whatever title', [
+            create_command('cmd-name', 'cmd help 1'),
+        ]),
+        ('test-group-2', 'other title', [
+            create_command('cmd-longer-name', 'cmd help 2'),
+        ]),
+    ]
+    generated_help = get_generated_help(groups)
+    expected_help = """
+        commands:
+
+            whatever title:
+                cmd-name          cmd help 1
+
+            other title:
+                cmd-longer-name   cmd help 2
+    """
+    assert generated_help == textwrap.dedent(expected_help)
+
+
+def test_customhelp_combined():
+    """Several commands in several groups."""
+    groups = [
+        ('test-group-1', 'this is very long title but we do not prejudge much', [
+            create_command('1-name', 'cmd help 1'),
+        ]),
+        ('test-group-2', 'short', [
+            create_command('cmd-B1', 'cmd help which is very long but whatever'),
+            create_command('cmd-B2', 'cmd help'),
+            create_command('cmd-longer-name', 'cmd help'),
+        ]),
+    ]
+    generated_help = get_generated_help(groups)
+    expected_help = """
+        commands:
+
+            this is very long title but we do not prejudge much:
+                1-name            cmd help 1
+
+            short:
+                cmd-B1            cmd help which is very long but whatever
```

Comment:
is it worth testing what happens if a commands help would want to wrap?
Where would we put the longer form help for a command? Right now we seem to only have 'summary'.

Reply:
This is just "the summary" of the command. It should be one short line.

We need to add support for each command to produce a long help text, which will be exposed only when getting the help of that specific command (`charmcraft foobar -h`), not here.

I opened #3 to discuss this.